Add a File ID mapping file and tests

cassella opened the pull request:
This is the same set of commits as the other PR, plus one more for consideration.

I added a cram test exercising the fileid mapping, an embedded fileid, and use of --relative-to.  I didn't want to build all the files "by hand" in the test, so I added them to the repo in tests/.  Now that I've enabled travis, I can tell the new test passes too.

I couldn't find any guidance as to whether this is a good approach.  Let me know if you'd like it done some other way (or not at all).

cassella committed:
Introduce fileid mapping file

This is an alternate source of the file<->File ID mapping.  It does
not interfere with the use of File IDs embedded in files.

This is meant to allow the use of file-specific dictionaries in a
project without requiring the addition of the scspell-id: tag into its
files.

Given an --override-dictionary=foo, the fileid mapping is stored in
foo.fileid.json, as a mapping from File ID to a list of filenames.  I
chose this arrangment (vs. filename->File ID) to make it easier to see
files that are sharing a fileid.

The filenames are stored as relative paths.  Absolute paths wouldn't
work very well for a dictionary checked into a project, and checked
out in different developer's working directories.  The directory
they're relative to is specified via the --relative-to commandline
argument.  (I expect that a project using this feature will have the
scspell invocation wrapped up in a script that'll provide the
--override-dictionary and --relative-to arguments so that this won't
be onerous.)

This has the disadvantage that the file-specific dictionary won't
"follow" the files accross renames and copies.  scspell.py options
will be introduced in subsequent commits to allow some basic
manipulations without editing the .json file.

I stored the mapping to a separate .json file primarily to take
advantage of json.dump/json.load dealing with spaces and newlines in
filenames.  As it happens, this leaves the regular dictionary file
backwards compatible with old versions of scspell.  scspell won't be
able to find the per-file dictionary for files specified there, but it
will otherwise be able to read and process the dictionary file.

Also change the generation of the handle_add() menu to generate each
optional menu item separately, instead of having a separate menu
hardcoded for each combination.

cassella committed:
Store fileid mappings sorted by filename

It's more reader-friendly and easier to edit this way.  And if the
file's being checked into source control, there'll be less churn.  (A
plain dict can be serialized in a completely different order each
time.)

Building the fileid map sorted by filename required a filename->id
dict, which I wanted for regular lookups anyway, so make that part of
the corpus object and maintain it on updates, rather than just
building it in place for the json.dumps().

cassella committed:
Break cmdline args up into groups

cassella committed:
Add cmdline arg --merge-fileids

cassella committed:
Add --rename-file command

cassella committed:
Add --delete-files command

cassella committed:
Indent all argparse.add_argument() lines the same way

cassella committed:
Swap order of --merge-fileids arguments to match --rename-file

cassella committed:
Move fileid generation into scspell/__init__.py

cassella committed:
README.rst: update for fileid mapping

cassella committed:
Change files from nargs='+' to nargs='*'

This way options that don't need the files list can be given without
needing extra "files" listed.  (e.g., --gen-id, and future ones)

cassella committed:
s/iteritems/items/ for python3

cassella committed:
Don't try to str.decode("utf-8") on python3

cassella committed:
Handle ValueError from json.load of fileid mapping

I saw this when an error writing the fileid mapping left it empty.

Handle (and report) the exception so spellchecking can continue.

cassella committed:
Add test.cram test for --fileidmap

cassella committed:
tests/fileidmap: new tests for fileid mapping manipulation

cassella committed:
corpus: add _fn_to_fq_rel() to convert filename to both fq and rel at once

cassella committed:
Fix --rename-file to respect relative filenames

cassella commented:
Rebased on top of the updated pr #5, and added tests of the fileid manipulation operations.

myint commented:
On my machine, `diff` doesn't have the option `--ignore-trailing-space`. A fix for that would be to avoid the trailing whitespace bug in Python 2.

``` diff
             try:
                 with io.open(mapping_file, mode='w', encoding='utf-8') as mf:
                     # http://stackoverflow.com/questions/36003023/json-dump-failing-with-must-be-unicode-not-str-typeerror
-                    tstr = json.dumps(od, ensure_ascii=False, indent=2)
+                    tstr = json.dumps(od, ensure_ascii=False,
+                                      indent=2, separators=(',', ': '))
```

myint commented on test.cram:
```diff
@@ -15,3 +15,38 @@ Test okay file.
+    > --relative-to tests/fileidmap \
+    > --delete-file tests/fileidmap/mix5.txt
+    Removing tests/fileidmap/mix5.txt <-> 7f72eb84-3576-11e6-a3a6-10ddb1d4c3d5 mappings
+    $ diff tests/fileidmap/dictionary tests/fileidmap/dictionary.post
```
I'm a bit confused about this `diff`. It looks like both these files are already in the repository. Are you just checking that the file hasn't changed?

cassella replied:
The repo versions are in $TESTDIR/tests.  Earlier in test.cram, I added "cp -a "$TESTDIR/tests ." to have a local copy to modify.  (and so invoking test.cram doesn't modify the git working directory.  (And to make all the commandlines shorter.))

The --merge-fileids and --rename-file and --delete-file invocations modify ./tests/fileidmap/dictionary.  Then the diff compares the result to ./tests/fileidmap/dictionary.post.  The .post file  is unmodified since the copy, and contains the expected result of those operations.  The diff could just as well have been
 `diff tests/fileidmap/dictionary $TESTDIR/tests/fileidmap/dictionary.post`

myint replied:
Ah right, I forgot about how cram works. It has been a while since I last used it.

myint commented:
By the way, if you don't mind, could you make sure the code still passes `pep8` and `pyflakes` after your changes.

```
$ pyflakes scspell*
scspell/_corpus.py:370: local variable 'rt_len' is assigned to but never used

$ pycodestyle scspell*
scspell/__init__.py:185:1: E302 expected 2 blank lines, found 1
scspell/__init__.py:276:1: E302 expected 2 blank lines, found 1
scspell/__init__.py:360:30: E261 at least two spaces before inline comment
scspell/__init__.py:541:28: E261 at least two spaces before inline comment
scspell/__init__.py:674:1: E302 expected 2 blank lines, found 1
scspell/__init__.py:694:1: E302 expected 2 blank lines, found 1
scspell/__init__.py:708:1: E302 expected 2 blank lines, found 1
scspell/__init__.py:716:1: E302 expected 2 blank lines, found 1
scspell/__init__.py:717:17: E128 continuation line under-indented for visual indent
scspell/_corpus.py:250:80: E501 line too long (89 > 79 characters)
scspell/_corpus.py:268:14: E231 missing whitespace after ','
scspell/_corpus.py:273:5: E303 too many blank lines (2)
scspell/_corpus.py:372:80: E501 line too long (80 > 79 characters)
scspell/_corpus.py:507:19: E703 statement ends with a semicolon
scspell.py:83:14: E251 unexpected spaces around keyword / parameter equals
```

cassella committed:
json.dumps using separators=(',', ': ')

This avoids trailing whitespace in dictionary.fileids.json files.

Update the tests/fileidmap/dictionary.fileids.json* to reflect the change.

cassella committed:
test.cram: invoke diff without --ignore-trailing-whitespace

This option isn't universally available.  The json.dumps change means
we don't need it.

cassella commented:
Those two take care of the diff issue.  I'll take care of the style issues.

Does "pass pep8" mean anything beyond a clean pycodestyle run?

cassella committed:
Remove unused rt_len variable

cassella committed:
Style/whitespace fixes

These now run clean:
  pyflakes scspell*
  pycodestyle scspell*

cassella committed:
test.cram: also remove comment about trailing whitespace

myint commented:
Thanks.
